### ISSUE TITLE
histc should (optionally) return a long tensor

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#32305]
It always returns integral values. Similar op `bincount` also returns a long tensor (when no `weights` is given).

**Comment by vadimkantorov** [comment]
(also discussed with @ngimel in https://github.com/pytorch/pytorch/issues/31475#issuecomment-567606676 )

**Comment by ezyang** [comment]
Yeah, it is pretty weird that bincount returns only float or double tensor.

**Comment by ghlai9665** [comment]
Why does histc return non-integer values? I'm getting values such as 7.6000e+01 in the array returned by `data.histc(40, 0, 10)`

**Comment by kurtamohler** [comment]
I imagine changing the return type is BC breaking and would require a deprecation cycle, right?

**Comment by vadimkantorov** [comment]
My proposal would be adding an explicit `dtype=` argument which would return int64 if it's set to int64 (but could also be set to uint8 or int16. also, the kernel could for small-sized inputs do accumulation into a smaller accumulation buffer if it can help). And during the deprecation cycle, it would first be a warning if it's not set explicitly, then an error if it's not set. Then at some distant point in the future, dtype=int64 could become the default and again not require `dtype=`

**Comment by vadimkantorov** [comment]
Also, I wonder if bincount / histc are faster than scatter_add

Hopefully they should be for uint8 inputs: https://github.com/pytorch/pytorch/issues/61819

Also, hopefully bincount / histc could support batched and multidim in the sense of doing histogram independently for some dimensions (parallelized by all other dimensions which would be considered batch dimensions)
